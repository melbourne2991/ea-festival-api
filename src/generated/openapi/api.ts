/* tslint:disable */
/* eslint-disable */
/**
 * Energy Australia Coding Test API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: v1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import { Configuration } from './configuration';
import globalAxios, { AxiosPromise, AxiosInstance, AxiosRequestConfig } from 'axios';
// Some imports not used depending on template conditions
// @ts-ignore
import { DUMMY_BASE_URL, assertParamExists, setApiKeyToObject, setBasicAuthToObject, setBearerAuthToObject, setOAuthToObject, setSearchParams, serializeDataIfNeeded, toPathString, createRequestFunction } from './common';
// @ts-ignore
import { BASE_PATH, COLLECTION_FORMATS, RequestArgs, BaseAPI, RequiredError } from './base';

/**
 * 
 * @export
 * @interface Band
 */
export interface Band {
    /**
     * 
     * @type {string}
     * @memberof Band
     */
    'name'?: string;
    /**
     * 
     * @type {string}
     * @memberof Band
     */
    'recordLabel'?: string;
}
/**
 * 
 * @export
 * @interface MusicFestival
 */
export interface MusicFestival {
    /**
     * 
     * @type {string}
     * @memberof MusicFestival
     */
    'name'?: string;
    /**
     * 
     * @type {Array<Band>}
     * @memberof MusicFestival
     */
    'bands'?: Array<Band>;
}

/**
 * FestivalsApi - axios parameter creator
 * @export
 */
export const FestivalsApiAxiosParamCreator = function (configuration?: Configuration) {
    return {
        /**
         * 
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        aPIFestivalsGet: async (options: AxiosRequestConfig = {}): Promise<RequestArgs> => {
            const localVarPath = `/api/v1/festivals`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }

            const localVarRequestOptions = { method: 'GET', ...baseOptions, ...options};
            const localVarHeaderParameter = {} as any;
            const localVarQueryParameter = {} as any;


    
            setSearchParams(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = {...localVarHeaderParameter, ...headersFromBaseOptions, ...options.headers};

            return {
                url: toPathString(localVarUrlObj),
                options: localVarRequestOptions,
            };
        },
    }
};

/**
 * FestivalsApi - functional programming interface
 * @export
 */
export const FestivalsApiFp = function(configuration?: Configuration) {
    const localVarAxiosParamCreator = FestivalsApiAxiosParamCreator(configuration)
    return {
        /**
         * 
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        async aPIFestivalsGet(options?: AxiosRequestConfig): Promise<(axios?: AxiosInstance, basePath?: string) => AxiosPromise<Array<MusicFestival>>> {
            const localVarAxiosArgs = await localVarAxiosParamCreator.aPIFestivalsGet(options);
            return createRequestFunction(localVarAxiosArgs, globalAxios, BASE_PATH, configuration);
        },
    }
};

/**
 * FestivalsApi - factory interface
 * @export
 */
export const FestivalsApiFactory = function (configuration?: Configuration, basePath?: string, axios?: AxiosInstance) {
    const localVarFp = FestivalsApiFp(configuration)
    return {
        /**
         * 
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        aPIFestivalsGet(options?: any): AxiosPromise<Array<MusicFestival>> {
            return localVarFp.aPIFestivalsGet(options).then((request) => request(axios, basePath));
        },
    };
};

/**
 * FestivalsApi - object-oriented interface
 * @export
 * @class FestivalsApi
 * @extends {BaseAPI}
 */
export class FestivalsApi extends BaseAPI {
    /**
     * 
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof FestivalsApi
     */
    public aPIFestivalsGet(options?: AxiosRequestConfig) {
        return FestivalsApiFp(this.configuration).aPIFestivalsGet(options).then((request) => request(this.axios, this.basePath));
    }
}


