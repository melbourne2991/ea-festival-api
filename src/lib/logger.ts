import bunyan from "bunyan";

class PrettyLogStream {
  write(chunk: object) {
    const log = chunk as { msg: string; time: Date; level: number };

    console.log(
      "[%s] %s: %s",
      log.time.toISOString(),
      bunyan.nameFromLevel[log.level],
      log.msg
    );
  }
}

export interface Logger {
  info(obj: Object, ...params: any[]): void
  warn(obj: Object, ...params: any[]): void
  debug(obj: Object, ...params: any[]): void
  error(obj: Object, ...params: any[]): void
  trace(obj: Object, ...params: any[]): void
}

export const logger = bunyan.createLogger({
  name: "ea-coding-test",
  streams: [
    {
      level: "info",
      stream: new PrettyLogStream(),
      type: "raw",
    },
  ],
});