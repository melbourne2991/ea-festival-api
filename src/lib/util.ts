export const requireEnvVar = (key: string): string => {
  const value = process.env[key]
  if (!value) throw new Error(`Missing env var: ${key}`)
  return value
};