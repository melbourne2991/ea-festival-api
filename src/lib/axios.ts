import axios from 'axios';
import axiosRetry from 'axios-retry';
import { logger } from './logger';

export const axiosInstance = axios.create({
  headers: {
    'Accept': 'application/json'
  }
});


const backoff = (retryCount: number) => 2 ** retryCount * 100

/**
 * Apply axios retry to our instance with exponential backoff for 429s
 */
axiosRetry(axiosInstance, {
  retries: 5,
  retryDelay: (count) => backoff(count),
  retryCondition: (error) => {
    return error.response?.status === 429
  },
  onRetry: (count, err) => {
    logger.warn(`Got ${err.response?.status} from API, retrying ... (${count})`)
  }
})