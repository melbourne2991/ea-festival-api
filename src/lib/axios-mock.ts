import {axiosInstance} from "./axios"
import MockAdapter from "axios-mock-adapter"
import { config } from "../config"

const mock = new MockAdapter(axiosInstance)

const scenarios = {
  basic: () => {
    mock.onGet(`${config.apiBaseUrl}/api/v1/festivals`).reply(200, [
      {
        name: "Festival e2e 1",
        bands: [
          {
            name: "Cool band",
            recordLabel: "That label",
          },
        ],
      },

      {
        name: "Festival e2e 2",
        bands: [
          {
            name: "Another band",
            recordLabel: "Another label",
          },
        ],
      },
    ], { 'Content-Type': 'application/json' })
  },

  'rate-limited': () => {
    mock
      .onGet(`${config.apiBaseUrl}/api/v1/festivals`)
      .replyOnce(429, 'Rate limited')
      .onGet(`${config.apiBaseUrl}/api/v1/festivals`)
      .replyOnce(429, 'Rate limited')
      .onGet(`${config.apiBaseUrl}/api/v1/festivals`)
      .replyOnce(200, [
        {
          name: "Festival e2e 1",
          bands: [
            {
              name: "Cool band",
              recordLabel: "That label",
            },
          ],
        },
  
        {
          name: "Festival e2e 2",
          bands: [
            {
              name: "Another band",
              recordLabel: "Another label",
            },
          ],
        },
      ])
  }
};

scenarios[config.testScenario! as keyof typeof scenarios]()

export { axiosInstance }