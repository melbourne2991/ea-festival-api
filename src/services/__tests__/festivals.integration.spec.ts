import { FestivalsApi, MusicFestival } from "../../generated/openapi";
import { Logger } from "../../lib/logger";
import { FestivalsService } from "../festivals.service";
import { PrinterService } from "../printer.service";

describe("Integration (Festival)", () => {
  let printerService: PrinterService;
  let festivalsApi: FestivalsApi;

  beforeEach(() => {
    jest.clearAllMocks();
    festivalsApi = new FestivalsApi();

    const logger = {
      info: () => {},
    } as unknown as Logger;

    const festivalsService = new FestivalsService(festivalsApi, logger);
    printerService = new PrinterService(festivalsService, logger);
  });

  it("should print out festival data", async () => {
    const festivals: MusicFestival[] = [
      {
        name: "Festival 1",
        bands: [
          {
            name: "Cool band",
            recordLabel: "That label",
          },
        ],
      },

      {
        name: "Festival 2",
        bands: [
          {
            name: "Another band",
            recordLabel: "Another label",
          },
        ],
      },
    ];

    jest
      .spyOn(festivalsApi, "aPIFestivalsGet")
      .mockImplementation()
      .mockResolvedValue({
        data: festivals,
      } as any);

    jest.spyOn(console, 'log')

    printerService.print();

    await new Promise(process.nextTick);

    expect((console.log as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  });
});
