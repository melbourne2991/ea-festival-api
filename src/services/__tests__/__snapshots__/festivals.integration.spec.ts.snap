// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration (Festival) should print out festival data 1`] = `
"That label
   Cool band
      Festival 1

Another label
   Another band
      Festival 2

"
`;
